core/lib/dal/src/contract_verification_dal.rs

```diff
@@ -560,6 +587,285 @@ impl ContractVerificationDal<'_, '_> {
+            )
+            .try_map(|row| {
+                let info = serde_json::from_value::<VerificationInfo>(row.verification_info)
+                    .decode_column("verification_info")?;
+                let bytecode_keccak256 = H256::from_slice(&row.bytecode_keccak256);
+                let bytecode_without_metadata_keccak256 = row
+                    .bytecode_without_metadata_keccak256
+                    .map(|h| H256::from_slice(&h));
+                Ok((
+                    info,
+                    bytecode_keccak256,
+                    bytecode_without_metadata_keccak256,
+                ))
+            })
+            .instrument("get_contract_verification_info_v2#perfect_match")
+            .with_arg("address", &address)
+            .fetch_optional(self.storage)
+            .await?
+        else {
+            return Ok(None);
+        };
+        if identifier.bytecode_keccak256 != bytecode_keccak256 {
+            // Sanity check
+            if bytecode_without_metadata_keccak256.is_none()
+                || identifier
+                    .bytecode_without_metadata_keccak256
+                    .map(|h| h.hash())
+                    != bytecode_without_metadata_keccak256
+            {
+                tracing::error!(
+                    contract_address = %address,
+                    identifier = ?identifier,
+                    bytecode_keccak256 = %bytecode_keccak256,
+                    info = ?info,
+                    "Bogus verification info fetched for contract",
+                );
+                anyhow::bail!("Internal error: bogus verification info detected");
+            }
+
+            // Mark the contract as partial match (regardless of other issues).
+            info.verification_problems = vec![VerificationProblem::IncorrectMetadata];
+        }
+        Ok(Some(info))
+    }
+
+    /// Checks if migration from `contracts_verification_info` to `contract_verification_info_v2` is performed
+    /// by checking if the latter has more or equal number of rows.
+    pub async fn is_verification_info_migration_performed(&mut self) -> DalResult<bool> {
+        let row = sqlx::query!(
+            r#"
+            SELECT
+                (SELECT COUNT(*) FROM contracts_verification_info) AS count_v1,
+                (SELECT COUNT(*) FROM contract_verification_info_v2) AS count_v2
+            "#,
+        )
+        .instrument("is_verification_info_migration_performed")
+        .fetch_one(self.storage)
+        .await?;
+
+        Ok(row.count_v2 >= row.count_v1)
+    }
+
+    pub async fn perform_verification_info_migration(
+        &mut self,
+        batch_size: usize,
+    ) -> anyhow::Result<()> {
+        // We use a long-running transaction, since the migration is one-time and during it
+        // no writes are expected to the tables, so locked rows are not a problem.
+        let mut transaction = self.storage.start_transaction().await?;
```

Comment:
I've decided to use a long-running transaction to make sure that we're fully idempotent (since we don't have IDs in the table, sorted addresses do not guarantee the correctness of resumed migration). Since no writes to either of tables are expected during migration, this should be OK.

Reply:
Yes; the migration is blocking and contract verifier is only run as a single instance, so in fact it's guaranteed that it won't happen.
I added a migration for the whole duration just in case something like this happens:
1. We deploy new contract verifier.
2. Something wrong happens during migration.
3. We rollback
4. New contracts are verified. 
5. We deploy new verifier again, and migration skips some rows because new entries were added _between_ runs.